File: api/src/main/java/net/blueshell/api/controller/BlogController.java
package net.blueshell.api.controller;

import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.annotation.security.PermitAll;
import jakarta.validation.Valid;
import lombok.extern.slf4j.Slf4j;
import net.blueshell.api.base.BaseController;
import net.blueshell.api.dto.BlogDTO;
import net.blueshell.api.mapper.BlogMapper;
import net.blueshell.api.service.BlogService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Slf4j
@RestController
@Tag(name = "Blogs")
public class BlogController extends BaseController<BlogService, BlogMapper> {

    @Autowired
    public BlogController(BlogService blogService, BlogMapper blogMapper) {
        super(blogService, blogMapper);
    }

    @PostMapping("/blogs")
    @PreAuthorize("hasAuthority('BOARD')")
    @ResponseStatus(HttpStatus.CREATED)
    public BlogDTO createBlog(@Valid @RequestBody BlogDTO dto) {
        var blog = mapper.fromDTO(dto);
        blog = service.create(blog);
        return mapper.toDTO(blog);
    }

    @PostMapping("/blogs/{id}")
    @PreAuthorize("hasAuthority('BOARD')")
    public BlogDTO updateBlog(@PathVariable("id") Long id, @Valid @RequestBody BlogDTO dto) {
        var blog = service.findById(id);
        mapper.fromDTO(dto, blog);
        blog = service.update(blog);
        return mapper.toDTO(blog);
    }

    @GetMapping("/blogs")
    @PermitAll
    public List<BlogDTO> findBlogs() {
        return mapper.toDTOs(service.findAll());
    }

    @GetMapping("/blogs/{id}")
    @PermitAll
    public BlogDTO findBlogById(@PathVariable Long id) {
        return mapper.toDTO(service.findById(id));
    }

    @DeleteMapping("/blogs/{id}")
    @PreAuthorize("hasAuthority('BOARD')")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteById(@PathVariable Long id) {
        service.deleteById(id);
    }
}

File: api/src/main/kotlin/net/blueshell/api/blog/web/BlogController.kt
package net.blueshell.api.blog.web

import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.annotation.security.PermitAll
import jakarta.validation.Valid
import net.blueshell.api.blog.web.dto.BlogDTO
import net.blueshell.api.blog.web.mapping.asEntity
import net.blueshell.api.blog.web.mapping.asDto
import net.blueshell.api.blog.application.BlogService
import net.blueshell.api.shared.web.BaseController
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpStatus
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@Tag(name = "Blogs")
class BlogController(blogService: BlogService) : BaseController<BlogService>(blogService) {
    @Value($$"${frontend.url}")
    private lateinit var frontendUrl: String

    @PostMapping("/blogs")
    @PreAuthorize("hasAuthority('BOARD')")
    @ResponseStatus(HttpStatus.CREATED)
    fun createBlog(@Valid @RequestBody dto: BlogDTO): BlogDTO {
        var blog = dto.asEntity()
        blog = service.create(blog)
        return blog.asDto(frontendUrl)
    }

    @PostMapping("/blogs/{id}")
    @PreAuthorize("hasAuthority('BOARD')")
    fun updateBlog(@PathVariable id: Long, @Valid @RequestBody dto: BlogDTO): BlogDTO {
        var blog = service.findById(id)
        dto.asEntity(blog)
        blog = service.update(blog)
        return blog.asDto(frontendUrl)
    }

    @GetMapping("/blogs")
    @PermitAll
    fun findBlogs(): MutableList<BlogDTO> {
        return service.findAll().map { it.asDto(frontendUrl) }.toMutableList()
    }

    @GetMapping("/blogs/{id}")
    @PermitAll
    fun findBlogById(@PathVariable id: Long): BlogDTO {
        return service.findById(id).asDto(frontendUrl)
    }

    @DeleteMapping("/blogs/{id}")
    @PreAuthorize("hasAuthority('BOARD')")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteById(@PathVariable id: Long) {
        service.deleteById(id)
    }
}
